R/write_pxl_file.R

```diff
@@ -0,0 +1,499 @@
+#' @param file A character string specifying the path to the .pxl
+#' file to be created.
+#' @param assay A character string specifying the name of the \code{CellGraphAssay5}.
+#' If set to \code{NULL} the default assay will be used.
+#' @param overwrite A logical value specifying whether to overwrite the \code{file}
+#' if it already exists.
+#'
+#' @return Nothing. The function writes the .pxl file to the specified location.
+#'
+#' @examples
+#' # Use Assay5 as the default assay version
+#' options(Seurat.object.assay.version = "v5")
+#'
+#' # Create Seurat object
+#' pxl_file <- system.file("extdata/five_cells",
+#'                         "five_cells.pxl",
+#'                         package = "pixelatorR")
+#' se <- ReadMPX_Seurat(pxl_file)
+#'
+#' se_merged <- merge(se, list(se, se, se))
+#' pxl_file <- fs::path_temp("small.pxl")
+#'
+#' # Export data to a new .pxl file
+#' WriteMPX_pxl_file(se_merged, pxl_file)
+#'
+#' # Read the new .pxl file
+#' se_merged <- ReadMPX_Seurat(pxl_file)
+#'
+#' # Reset global option
+#' options(Seurat.object.assay.version = "v3")
+#'
+#' @export
+#'
+WriteMPX_pxl_file <- function (
+  object,
+  file,
+  assay = NULL,
+  overwrite = FALSE
+) {
+
+  stopifnot(
+    "'object' must be a Seurat object" =
+      inherits(object, what = "Seurat"),
+    "'file' must be a valid path ending with .pxl" =
+      is.character(file) && (fs::path_ext(file) == "pxl"),
+    "'overwrite' must be either TRUE or FALSE" =
+      is.logical(overwrite)
+  )
+
+  # Check if file exists
+  if (fs::file_exists(file) & !overwrite) {
+    abort(glue("{col_br_blue(file)} already exists. Please select a different ",
+               "file name or set overwrite = TRUE if you are certain that the ",
+               "existing file should be replaced."))
+  }
+
+  assay <- assay %||% DefaultAssay(object)
+  cg_assay <- object[[assay]]
+  stopifnot(
+    "'assay' must be a 'CellGraphAssay' or a 'CellGraphAssay5' object" =
+      is(cg_assay, "MPXAssay")
+  )
+
+  # fetch and validate fs_map
+  fs_map <- cg_assay@fs_map
+  .validate_fs_map(fs_map)
+
+  # Delete the file only if it's not in use
+  # by the current object
+  if (fs::file_exists(file)) {
+    if (file %in% fs_map$pxl_file) {
+      abort(glue(
+        "The selected file name '{col_br_blue(file)}' is currently in use ",
+        "by the input object. You need to select a different file name."
+      ))
+    }
+    fs::file_delete(file)
+  }
+
+  # Create a temporary directory
+  pxl_folder <- file.path(fs::path_temp(), .generate_random_string())
+  fs::dir_create(path = pxl_folder)
+
+  cli_rule(left = "Creating .pxl file")
+
+  # Export spatial metrics
+  .write_spatial_metrics_to_parquet(object = object, pxl_folder = pxl_folder)
+
+  # Update and write edgelist data
+  .merge_and_write_edgelists_to_parquet(fs_map = fs_map, pxl_folder = pxl_folder)
+
+  # Extract and merge meta data
+  .write_json_meta_data(fs_map = fs_map, pxl_folder = pxl_folder)
```

Comment:
Looking at this function, it looks like the json file is generated from a list of strings from the original json files, if there are multiple. Is that correct? How does that work with `pixelator`?

Reply:
The metadata.json files are parsed in R with jsonlite which returns the content as a list object. I've tried to mimic the `pixelator` behaviour by adding a top level field called "sample". When the merged list is exported with jsonlite, the list structure is translated in to json format. This is what it could look like after merging:

````
{
    "samples": {
        "sample1": {
            "version": "0.12.0",
            "sample": "Sample07_SG_CD3_S7_001",
            "file_format_version": 1,
            "analysis": {
                "params": {
                    "compute_polarization": true,
                    "compute_colocalization": true,
                    "use_full_bipartite": false,
                    "polarization_normalization": "clr",
                    "polarization_binarization": false,
                    "colocalization_transformation": "log1p",
                    "colocalization_neighbourhood_size": 1,
                    "colocalization_n_permutations": 50,
                    "colocalization_min_region_count": 5
                }
            }
        },
        "sample2": {
            "version": "0.12.0",
            "sample": "Sample02_SG_unstimulated_S2_001",
            "file_format_version": 1,
            "analysis": {
                "params": {
                    "compute_polarization": true,
                    "compute_colocalization": true,
                    "use_full_bipartite": false,
                    "polarization_normalization": "clr",
                    "polarization_binarization": false,
                    "colocalization_transformation": "log1p",
                    "colocalization_neighbourhood_size": 1,
                    "colocalization_n_permutations": 50,
                    "colocalization_min_region_count": 5
                }
            }
        },
        "sample3": {
            "version": "0.12.0",
            "sample": "Uropod_CD54_Rantes_coated_S3_001",
            "file_format_version": 1,
            "analysis": {
                "params": {
                    "compute_polarization": true,
                    "compute_colocalization": true,
                    "use_full_bipartite": false,
                    "polarization_normalization": "clr",
                    "polarization_binarization": false,
                    "colocalization_transformation": "log1p",
                    "colocalization_neighbourhood_size": 1,
                    "colocalization_n_permutations": 50,
                    "colocalization_min_region_count": 5
                }
            }
        }
    }
}
````

But thanks for pointing this out. I need to double check that the output look as expected.

Reply:
Ok so there were some issue with the merging which should be resolved now! "samples" was called "sample" and I added `auto_unbox` to unbox all atomic vectors of length 1. Without this, all vectors in the json file are formatted as arrrays. For instance, without it, it would look something like this:

````
"colocalization_n_permutations": [50]
````

instead of:
````
"colocalization_n_permutations": 50
````

I don't think it should matter, but adding this makes the file more readable.